static/app/components/notificationActions/notificationActionItem.tsx

```diff
@@ -0,0 +1,344 @@
+  const [isEditing, setIsEditing] = useState(defaultEdit);
+  const [updatedAction, setUpdatedAction] = useState(action);
+  const serviceType = action.serviceType;
+  const api = useApi();
+  const organization = useOrganization();
+
+  function renderIcon() {
+    // Currently email and Sentry notification use the same icon
+    if (
+      serviceType === NotificationActionService.EMAIL ||
+      serviceType === NotificationActionService.SENTRY_NOTIFICATION
+    ) {
+      return <IconMail size="sm" />;
+    }
+    return <PluginIcon pluginId={serviceType} size={16} />;
+  }
+
+  function renderDescription() {
+    // TODO(enterprise): descriptions for email, msteams, sentry_app
+    if (serviceType === NotificationActionService.SENTRY_NOTIFICATION) {
+      const roleElements = recipientRoles?.map(role => (
+        <NotificationRecipient key={role}>{role}</NotificationRecipient>
+      ));
+      return (
+        <Fragment>
+          <div>{t('Send an email notification to the following roles')}</div>
+          {roleElements}
+        </Fragment>
+      );
+    }
+    return (
+      <Fragment>
+        <div>{t('Send a notification to the')}</div>
+        <NotificationRecipient>{action.targetDisplay}</NotificationRecipient>
+        <div>
+          {action.serviceType === NotificationActionService.SLACK
+            ? t('channel')
+            : t('service')}
+        </div>
+      </Fragment>
+    );
+  }
+
+  async function handleDelete() {
+    const endpoint = `/organizations/${organization.slug}/notifications/actions/${action.id}/`;
+    try {
+      await api.requestPromise(endpoint, {
+        method: 'DELETE',
+      });
+      addSuccessMessage(t('Successfully deleted notification action'));
+      onDelete(index);
+    } catch (err) {
+      addErrorMessage(t('Unable to delete notification action'));
+    }
+  }
+
+  function handleCancel() {
+    if (action.id) {
+      setIsEditing(false);
+      return;
+    }
+    // Delete the unsaved notification action
+    onDelete(index);
+  }
+
+  async function handleSave() {
+    const formProps = getFormProps();
+    addLoadingMessage();
+    // TODO(enterprise): use "requires" to get data to send
+    // This is currently optimized for spike protection
+    const data = cloneDeep(updatedAction);
+
+    // Remove keys from the data if they are falsy
+    Object.keys(data).forEach(key => {
+      if (!data[key]) {
+        delete data[key];
+      }
+    });
+
+    try {
+      const resp = await api.requestPromise(formProps.apiEndpoint, {
+        method: formProps.apiMethod,
+        data: {...data, projects: [project.slug]},
+      });
+      addSuccessMessage(formProps.successMessage);
+      onUpdate(index, resp);
+      setUpdatedAction(resp);
+      setIsEditing(false);
+    } catch (err) {
+      addErrorMessage(formProps.errorMessage);
+    }
+  }
+
+  function handleChange(name: string, value: any) {
+    const modifiedAction = cloneDeep(updatedAction);
```

Comment:
That's not correct, she'll need to change the identity of the `updatedAction` object (as @leeandher has recommended below) in order for react to understand that the value has changed.

`cloneDeep` works here because it produces a new identity of the `updatedAction`, so when calling `setUpdatedAction(modifiedAction);` react will re-render. If you simply mutate a value inside of the object **the identity of the object will be the same, and react will not re-render**. [The react documentation covers this well](https://react.dev/learn/updating-objects-in-state)

> But you shouldn’t change objects that you hold in the React state directly. Instead, when you want to update an object, you need to create a new one (or make a copy of an existing one), and then set the state to use that copy.

Reply:
to be clear, @leeandher is correct, you do not need cloneDeep here, a shallow clone `{...updatedAction, [name]: value}` is enough to create a new object

Reply:
Ah - yes and yes